rfcs/0007-synchronous-commit.md

```diff
@@ -0,0 +1,446 @@
+
+In conclusion, the difference between the current model and this proposal is adding a "Committed" watermark. And as it does not make sense to use `DurabilityLevel` to contain it, we propose to use `ReadWatermark` to identify the positions of the commit history for read operations.
+
+### Sync Commit
+
+In contrast to the read side which may mix the positions in term of "durability" and "commit", the write / commit side is more straightforward, it only need to track the things about durability.
+
+To implement Synchronous Commit semantics, the write side remains do not need to change at all: the operation returns once the WAL is persisted according to the specified durability level. A `DurabilityLevel` is all we need here.
+
+Like:
+
+```rust
+let opts = WriteOptions::new().with_await_durability(DurabilityLevel::Remote);
+db.write(key, value, opts).await?;
+```
+
+In other systems, this option is often named as `sync`, to emphasize the idea of synchronous commit. To respect the idea of Least Surprise, it might be not a bad idea to rename `await_durability` to `sync`.
+
+```rust
+let opts = WriteOptions::new().with_sync(DurabilityLevel::Memory);
+db.write(key, value, opts).await?;
+```
+
+However, for the writes that not expected to durablely persisted, this can be thought of as **disabling** synchronous commit entirely rather than synchronizing to memory. The code sample may be like:
+
+```rust
+enum SyncLevel {
+    Off,
+    Local,
+    Remote,
+}
+
+let opts = WriteOptions::new().with_sync(SyncLevel::Off);
+db.write(key, value, opts).await?;
+```
+
+The benefits of this naming change are:
+
+1. The name "Synchronous Commit" directly describes the feature we aim to support.
+2. The term "sync" is commonly used across database systems and is well understood, familiar to users.
+3. The names are shorter, easier to type & read.
+4. It's accurate to describe as sync commit as "off" when it comes with memory durability.
+
+The inner implementation of the write side do not need to change, it's still just await the data to be persisted to storage as the specified durability level, then return. All we put into discussion here is the naming stuff.
+
+## Implementation
+
+The key change is to seperate "Append to WAL" and "Commit to make it visible" into two different steps.
+
+In the current design, whenever a write operation is called, it'll append the data to the WAL, the data is becoming visible to readers with `DurabilityLevel::Memory` immediately, as the readers will read the data from the WAL in front of MemTable. And when the WAL buffer reaches the `flush.interval` or `flush.size`, it'll be flushed to storage, and apply to MemTable.
+
+The sequence diagram is like this:
+
+```mermaid
+sequenceDiagram
+    participant WriteBatch
+    participant WAL
+    participant MemTable
+    participant Flusher
+
+    WriteBatch->>WAL: append write op to current WAL
```

Comment:
Can't remember if we read from the current wal or just `imm_wals` in `db_state`. But I think it's irrelevant for this discussion since `imm_wals` are also in-memory only.

Reply:
yes, the WAL and flusher in this diagram are logical, and they may differ from the actual implementation. however, i suppose the conceptual difference might not be significant.